cpp/src/parquet/column_reader.cc

```diff
@@ -1115,20 +1129,27 @@ int64_t TypedColumnReaderImpl<DType>::ReadBatch(int64_t batch_size, int16_t* def
   // TODO(wesm): keep reading data pages until batch_size is reached, or the
   // row group is finished
   int64_t num_def_levels = 0;
-  int64_t values_to_read = 0;
-  ReadLevels(batch_size, def_levels, rep_levels, &num_def_levels, &values_to_read);
-
-  *values_read = this->ReadValues(values_to_read, values);
+  // Number of non-null values to read within `num_def_levels`.
+  int64_t non_null_values_to_read = 0;
+  ReadLevels(batch_size, def_levels, rep_levels, &num_def_levels,
+             &non_null_values_to_read);
+  // Should not return more values than available in the current data page.
+  ARROW_DCHECK_LE(num_def_levels, this->available_values_current_page());
```

Comment:
shouldn't this throw?  it's not clear if this can be triggered by bad data?

Reply:
IMO, throw is the duty of `ReadLevels`, a internal `ReadLevel` implementation cannot read more data than input batch-size

Reply:
I think the question is: can a corrupt Parquet file trigger this error?

Reply:
Never, since:

```c++
// ReadLevels will throw exception when any num-levels read is not equal to the number
// of the levels can be read.
```

Here we just do dcheck to check that `ReadLevels` has well done this

Reply:
I guess if we are concerned with it, it might pay to still check it and throw an exception in case there is in fact a bug, I don't think this code is on the inner loop (it would be nice to have a macro that does this to avoid code bloat.

Reply:
done. About macro I think it's just change three line to two, and may hack for func call...so keep it here